### PR TITLE
Fixed #28560 -- Fixed ordered queries using the values()/values_list() with distinct().

### PR DESCRIPTION
**Pull request by felixxm**
https://code.djangoproject.com/ticket/28560

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -539,13 +539,14 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
 
-            if self.query.subquery and extra_select:
-                # If the query is used as a subquery, the extra selects would
-                # result in more columns than the left-hand side expression is
-                # expecting. This can happen when a subquery uses a combination
-                # of order_by() and distinct(), forcing the ordering expressions
-                # to be selected as well. Wrap the query in another subquery
-                # to exclude extraneous selects.
+            if (self.query.subquery or self.query.values_select) and extra_select:
+                # If the query is used as a subquery or has limited list of
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: has a limited

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -539,13 +539,14 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                # columns, the extra selects would result in more columns than
+                # the left-hand side expression is expecting. This can happen
+                # when a subquery uses a combination of order_by(), values() or
+                # values_list() and distinct(), forcing the ordering
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: , and

**Comment by felixxm** [comment]
@timgraham Thanks for review. I updated PR and I restored `test_extra_values_distinct_ordering`, but it seems that we cannot preserve ordering from subquery 😕, therefore it won't work as expected in some cases.

**Comment by timgraham** [comment]
I'm not sure what that means for this issue. Is it only an issue with `extra()`? If so, that's probably fine since it's discouraged and likely to be deprecated.

**Comment by felixxm** [comment]
This failure is not strictly related with `extra()`. If we wrap the query in a subquery we will solve the issue described in the ticket but we cannot preserve ordering e.g.
```sql
SELECT DISTINCT subquery.column_a
FROM (
    SELECT column_a, column_b FROM table ORDER BY column_b
) subquery
```
for a table
```
column_a | column_b
-------------------
3        |       3 
3        |       2
4        |       1
```
may return `3, 4` or `4, 3`, but without a subquery it will always return `4, 3, 3`. On the other hand for a table without duplicate values (like in `test_extra_values_distinct_ordering`) e.g. for a table
```
column_a | column_b
-------------------
3        |       3 
4        |       1
```
it may return `3, 4` or `4, 3`, but without a subquery it will always return `4, 3`. :face_with_head_bandage:  \cc @charettes

**Comment by felixxm** [comment]
If anyone would like to continue working on this patch, please feel free.

**Pull request closed without merging**